File: src/2025-02-27-deepclaude37-rust/java/exercises/practice/bank-account/src/main/java/BankAccount.java
class BankAccount {
    private volatile boolean isOpen;
    private int balance;
    
    public BankAccount() {
        this.isOpen = false;
        this.balance = 0;
    }

    synchronized void open() throws BankAccountActionInvalidException {
        if (isOpen) {
            throw new BankAccountActionInvalidException("Account is already open");
        }
        isOpen = true;
        balance = 0;
    }

    synchronized void close() throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is not open");
        }
        isOpen = false;
    }

    synchronized int getBalance() throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        return balance;
    }

    synchronized void deposit(int amount) throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        if (amount <= 0) {
            throw new BankAccountActionInvalidException("Cannot deposit non-positive amount");
        }
        balance += amount;
    }

    synchronized void withdraw(int amount) throws BankAccountActionInvalidException {
        if (!isOpen) {
            throw new BankAccountActionInvalidException("Account is closed");
        }
        if (amount <= 0) {
            throw new BankAccountActionInvalidException("Cannot withdraw non-positive amount");
        }
        if (balance < amount) {
            throw new BankAccountActionInvalidException("Cannot withdraw more than balance");
        }
        balance -= amount;
    }

}

File: src/2025-02-27-deepclaude37-rust/cpp/exercises/practice/bank-account/bank_account.cpp
#include "bank_account.h"
#include <stdexcept>

namespace Bankaccount {

Bankaccount::Bankaccount() : balance(0.0), is_closed(true) {}

void Bankaccount::open() {
    std::lock_guard<std::mutex> lock(mtx);
    if (!is_closed) {
        throw std::runtime_error("Account already open");
    }
    is_closed = false;
    balance = 0.0;
}

void Bankaccount::deposit(double amount) {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account is closed");
    }
    if (amount <= 0) {
        throw std::runtime_error("Invalid deposit amount");
    }
    balance += amount;
}

void Bankaccount::withdraw(double amount) {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account is closed");
    }
    if (amount <= 0) {
        throw std::runtime_error("Invalid withdrawal amount");
    }
    if (amount > balance) {
        throw std::runtime_error("Insufficient funds");
    }
    balance -= amount;
}

void Bankaccount::close() {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account already closed");
    }
    is_closed = true;
}

double Bankaccount::balance() const {
    std::lock_guard<std::mutex> lock(mtx);
    if (is_closed) {
        throw std::runtime_error("Account is closed");
    }
    return balance;
}

}  // namespace Bankaccount
